core/code/chat.js

```diff
@@ -339,158 +342,227 @@ window.chat.writeDataToHash = function(newData, storageHash, isPublicChannel, is
+  var nick = '';
+  markup.forEach(function(ent) {
+    switch (ent[0]) {
+    case 'SENDER': // user generated messages
+      nick = ent[1].plain.slice(0, -2); // cut “: ” at end
+      break;
 
+    case 'PLAYER': // automatically generated messages
+      nick = ent[1].plain;
+      team = ent[1].team === 'RESISTANCE' ? TEAM_RES : TEAM_ENL;
+      break;
 
-    // format: timestamp, autogenerated, HTML message
-    storageHash.data[json[0]] = [json[1], auto, chat.renderMsg(msg, nick, time, team, msgToPlayer, systemNarrowcast), nick];
+    default:
+      break;
+    }
+  });
 
+  return {
+    guid: data[0],
+    time: time,
+    public: isPublic,
+    secure: isSecure,
+    alert: msgAlert,
+    msgToPlayer: msgToPlayer,
+    type: data[2].plext.plextType,
+    narrowcast: systemNarrowcast,
+    auto: auto,
+    player: {
+      name: nick,
+      team: team,
+    },
+    markup: markup,
+  };
+};
+
+window.chat.writeDataToHash = function(newData, storageHash, isPublicChannel, isOlderMsgs, isAscendingOrder) {
+  window.chat.updateOldNewHash(newData, storageHash, isOlderMsgs, isAscendingOrder);
+
+  newData.result.forEach(function(json) {
+    // avoid duplicates
+    if (json[0] in storageHash.data) return true;
+
+    var parsedData = chat.parseMsgData(json);
+
+    // format: timestamp, autogenerated, HTML message, nick, additional data (parsed, plugin specific data...)
+    storageHash.data[parsedData.guid] = [parsedData.time, parsedData.auto, chat.renderMsgRow(parsedData), parsedData.player.name, parsedData];
+    if (isAscendingOrder) {
+      storageHash.guids.push(parsedData.guid);
+    } else {
+      storageHash.guids.unshift(parsedData.guid);
+    }
   });
-}
+};
+
+//
+// Rendering primitive for markup, chat cells (td) and chat row (tr)
+//
+
+window.chat.renderText = function (text) {
+  return $('<div>').text(text.plain).html().autoLink();
+};
 
 // Override portal names that are used over and over, such as 'US Post Office'
 window.chat.getChatPortalName = function(markup) {
   var name = markup.name;
-  if(name === 'US Post Office') {
+  if (name === 'US Post Office') {
     var address = markup.address.split(',');
     name = 'USPS: ' + address[0];
   }
   return name;
-}
+};
 
-// renders data from the data-hash to the element defined by the given
-// ID. Set 3rd argument to true if it is likely that old data has been
-// added. Latter is only required for scrolling.
-window.chat.renderData = function(data, element, likelyWereOldMsgs) {
-  var elm = $('#'+element);
-  if(elm.is(':hidden')) return;
+window.chat.renderPortal = function (portal) {
+  var lat = portal.latE6/1E6, lng = portal.lngE6/1E6;
+  var perma = window.makePermalink([lat,lng]);
+  var js = 'window.selectPortalByLatLng('+lat+', '+lng+');return false';
+  return '<a onclick="'+js+'"'
+    + ' title="'+portal.address+'"'
+    + ' href="'+perma+'" class="help">'
+    + window.chat.getChatPortalName(portal)
+    + '</a>';
```

Comment:
Can we get rid of 'onclick' attribute?
Here and everywhere.

Reply:
how to jump to a portal then ? chat is built by a huge raw html string that cannot hold htmlelement event.
Each row (comm log) is calculated once and store in the channel data. We store a string, it could be a node but that'd be another step for the UI.

Reply:
> We store a string, it could be a node

We could store jQuery object instead. But it's out of scope of this PR, I agree.

Reply:
are jQuery object lighter than html elements ?

Reply:
Sure not lighter.
I just say that current 'string' approach is not good at all